src/alire/alire-test_runner.adb

```diff
@@ -0,0 +1,232 @@
+
+      function Fail_Count return Natural;
+      --  Get the number of failed tests
+   private
+      Passed : Natural := 0;
+      Failed : Natural := 0;
+   end Driver;
+
+   protected body Driver is
+      procedure Pass (Msg : String) is
+      begin
+         Passed := Passed + 1;
+         Trace.Always ("[ " & CLIC.TTY.OK ("PASS") & " ] " & Msg);
+      end Pass;
+
+      procedure Fail (Msg : String; Output : AAA.Strings.Vector) is
+      begin
+         Failed := Failed + 1;
+         Trace.Always ("[ " & CLIC.TTY.Error ("FAIL") & " ] " & Msg);
+         if not Output.Is_Empty then
+            Trace.Always ("*** Test output ***");
+            for L of Output loop
+               Trace.Always (CLIC.TTY.Dim (L));
+            end loop;
+            Trace.Always ("*** End Test output ***");
+         end if;
+      end Fail;
+
+      function Total_Count return Natural
+      is (Passed + Failed);
+      function Fail_Count return Natural
+      is (Failed);
+   end Driver;
+
+   procedure Create_Gpr_List
+     (Root : Alire.Roots.Root; List : AAA.Strings.Vector)
+     --  Create a gpr file containing a list of the test files
+     --  (named `Test_Files`).
+
+   is
+      File_Path : constant Alire.Absolute_Path :=
+        Root.Path / "config" / (Root.Name.As_String & "_list_config.gpr");
+      File      : Text_Files.File := Text_Files.Create (File_Path);
+      Lines     : access AAA.Strings.Vector renames File.Lines;
+      First     : Boolean := True;
+
+      Indent : constant String := "   ";
+
+      Root_Name : constant String :=
+        AAA.Strings.To_Mixed_Case (Root.Name.As_String);
+   begin
+      Touch (File_Path, True);
+
+      Lines.Append_Line ("abstract project " & Root_Name & "_List_Config is");
+      Lines.Append_Line (Indent & "Test_Files := (");
+
+      for Name of List loop
+         Lines.Append_Line (Indent & Indent);
+         if First then
+            Lines.Append_To_Last_Line (" ");
+            First := False;
+         else
+            Lines.Append_To_Last_Line (",");
+         end if;
+         Lines.Append_To_Last_Line ("""" & Name & ".adb""");
+      end loop;
+
+      Lines.Append_Line (Indent & ");");
+      Lines.Append_Line ("end " & Root_Name & "_List_Config;");
+   end Create_Gpr_List;
+
+   procedure Run_All_Tests
+     (Root : Alire.Roots.Root; Test_List : AAA.Strings.Vector; Jobs : Positive)
+   is
+      use GNAT.OS_Lib;
+
+      function Cmp (A, B : Process_Id) return Boolean
+      is (Pid_To_Integer (A) < Pid_To_Integer (B));
+
+      package Map is new
+        Ada.Containers.Indefinite_Ordered_Maps
+          (Process_Id,
+           String,
+           "<" => Cmp);
+
+      Running_Tests : Map.Map := Map.Empty_Map;
+      Output_Files  : Map.Map := Map.Empty_Map;
+
+      procedure Spawn_Test (Test_Name : String) is
+         Exe_Name : constant String := Test_Name & Alire.OS_Lib.Exe_Suffix;
+         Filename : constant String := "output_" & Test_Name & ".tmp";
```

Comment:
This is going in the current working directory?

Reply:
```suggestion
         Filename : constant String :=
            Root.Working_Folder / ("output_" & Test_Name & ".tmp");
```

@mosteo is the `alire` folder acceptable for temporary test output files? If nothing goes catastrophically wrong they should always be cleaned up.

Reply:
Yes, it's actually the [preferred location for our temp](https://github.com/alire-project/alire/blob/f5c4bb3fb4116b1b88bd564d7b3a64e7fc981e64/src/alire/alire-directories.adb#L600) files when it is locatable (actually we use `alire/tmp`).